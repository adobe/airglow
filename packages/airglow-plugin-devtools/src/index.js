/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import React from 'react';
import { POST_ENHANCER, COMPONENT, COMPOSER } from 'airglow';
import Devtools from './devtools';

export default function devtoolsPlugin({
  reduxDevtools = false,
  reduxBrowser = false
} = {}) {
  return (engine) => {
    if (reduxDevtools) {
      engine.plugin(POST_ENHANCER, Devtools.instrument());
      engine.plugin(COMPONENT, <Devtools key="c" />);
    }
    if (reduxBrowser) {
      engine.plugin(COMPOSER, browserCompose);
    }
  };
}

/* eslint-disable no-underscore-dangle */
const browserHook = () => global.window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__;

const browserCompose = compose =>
  (browserHook() ? browserHook()({}) : compose);
