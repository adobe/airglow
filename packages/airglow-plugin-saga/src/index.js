/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import createSagaMiddleware from 'redux-saga';
import { FEATURE, MIDDLEWARE, BOOTSTRAP_MODULE } from 'airglow';

export const SAGA = 'wire_sagas';

export default function sagaPlugin() {
  const middleware = createSagaMiddleware();

  return (engine) => {
    engine.plugin(MIDDLEWARE, middleware);
    engine.plugin(FEATURE, sagaFeature(middleware));
    engine.plugin(BOOTSTRAP_MODULE, bootstrapSagas(middleware));
  };
}

const sagaFeature = middleware => lookup =>
  lookup(SAGA).forEach(
    plugin => middleware.run(plugin)
  );

const bootstrapSagas = middleware => ({ sagas }) => {
  if (!sagas) { return; }
  const runSagas = Array.isArray(sagas) ? sagas : [sagas];
  runSagas.forEach(saga => middleware.run(saga));
};
