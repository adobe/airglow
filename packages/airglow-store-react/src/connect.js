/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import React from 'react';
import { StoreConsumer } from './store.context';

export default (mapState = R.identity, mapHandlers = R.identity) => Container => props => (
  <StoreConsumer>
    {({ state, store }) => (
      <Container
        {...R.mergeDeepRight(
          props,
          R.mergeDeepRight(
            mapState(state),
            mapHandlers(store.dispatch)
          )
        )}
      />
    )}
  </StoreConsumer>
);
