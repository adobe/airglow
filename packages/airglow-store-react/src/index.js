/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';

import { REDUCER, HOC } from 'airglow';
import { StoreProvider } from './store.context';

export { default as connect } from './connect';

let SUBSCRIBE_INDEX = 0;

class Store {
  constructor(reducers = []) {
    this.state = {};
    this.reducers = reducers;
    this.dispatch = this.dispatch.bind(this);
    this.injectReducers = this.injectReducers.bind(this);
    this.subscriptions = {};
    this.dispatch({ type: 'REACT.STORE.INITIALIZE' });
  }

  dispatch(action) {
    this.state = R.reduce(
      (newState, reducer) => reducer(newState, action),
      this.state,
      this.reducers || []
    );
    for (let i = 0; i <= SUBSCRIBE_INDEX; ++i) {
      if (this.subscriptions[i]) {
        this.subscriptions[i](this);
      }
    }
  }

  getState() {
    return this.state;
  }

  subscribe(callback) {
    this.subscriptions[++SUBSCRIBE_INDEX] = callback;
    return () => { this.subscriptions[SUBSCRIBE_INDEX] = null; };
  }

  injectReducers(newReducers) {
    if (!newReducers) { return; }
    this.reducers = R.concat(this.reducers, newReducers);
    this.dispatch({ type: 'REACT.STORE.INJECT.REDUCERS' });
  }
}

export const createStore = reducers => new Store(reducers);

class StoreFactory {
  create(lookup) {
    this.store = createStore(getReducers(lookup));
    return this.store;
  }

  bootstrap(engine) {
    engine.plugin(HOC, {
      component: StoreProvider,
      props: {
        store: this.store
      }
    });
  }
}

const getReducers = lookup => reduceLookup(lookup(REDUCER));

const reduceLookup = data => data.reduce(
  (results, plugin) => mergeInto(results, plugin), []
);

const mergeInto = (source, toMerge) =>
  source.concat(Array.isArray(toMerge) ? toMerge : [toMerge]);

export default () => new StoreFactory();
