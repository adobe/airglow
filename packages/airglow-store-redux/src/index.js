/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { reduce } from 'ramda';
import { compose, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import { REDUCER, POST_ENHANCER, ENHANCER, MIDDLEWARE, COMPOSER, HOC } from 'airglow';
import lazyStore from './lazy.store';

export { default as connect } from './connect';

class StoreFactory {
  create(lookup) {
    this.store = lazyStore(getReducers(lookup), getEnhancers(lookup));
    return this.store;
  }

  bootstrap(engine) {
    engine.plugin(HOC, {
      component: Provider,
      props: {
        store: this.store
      }
    });
  }
}

const getReducers = lookup => reduceLookup(lookup(REDUCER));
const getEnhancers = lookup =>
  getComposer(lookup)(
    ...getPreEnhancers(lookup),
    applyMiddleware(...getMiddleware(lookup)),
    ...getPostEnhancers(lookup)
  );
const getPreEnhancers = lookup => reduceLookup(lookup(ENHANCER));
const getPostEnhancers = lookup => reduceLookup(lookup(POST_ENHANCER));
const getMiddleware = lookup => reduceLookup(lookup(MIDDLEWARE));

const getComposer = lookup =>
  reduce(
    (acc, custom) => custom(acc),
    compose,
    lookup(COMPOSER)
  );

const reduceLookup = data => data.reduce(
  (results, plugin) => mergeInto(results, plugin), []
);

const mergeInto = (source, toMerge) =>
  source.concat(Array.isArray(toMerge) ? toMerge : [toMerge]);

export default () => new StoreFactory();
