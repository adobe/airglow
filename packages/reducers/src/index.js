/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import slice from './reducers/slice.reducer';
import call from './reducers/call.reducer';
import copy from './reducers/copy.reducer';
import toggle from './reducers/toggle.reducer';
import initialize from './reducers/initialize.reducer';
import reset from './reducers/reset.reducer';
import reduce from './reducers/reduce.reducer';
import extract from './reducers/extract.reducer';

import createAction from './create.action';

export {
  slice,
  call,
  copy,
  toggle,
  reset,
  extract,
  initialize,
  reduce,
  createAction
};
