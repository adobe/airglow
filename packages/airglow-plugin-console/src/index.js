/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import { REDUCER, BOOTSTRAP_MODULE, BOOTSTRAP_PLUGIN } from 'airglow';
import { slice, createAction } from '@airglow/reducers';
import { BOOTSTRAP_CONSOLE } from './actions';
import reducer from './reducers';

const STORE_KEY = 'console-commands';
const doBootstrap  = createAction(BOOTSTRAP_CONSOLE);

export default function prefabPlugin(config) {
  return (engine) => {
    engine.plugin(REDUCER, slice(STORE_KEY).with(reducer));
    engine.plugin(BOOTSTRAP_MODULE, bootstrap(engine));
    engine.plugin(BOOTSTRAP_PLUGIN, bootstrapPlugin(engine, config));
  };
}

const logHelp = ({ state }) => {
  R.forEachObjIndexed(
    ({ description }, key) => {
      console.info(`${key}(): ${description}`); // eslint-disable-line no-console
    },
    state['console-commands']
  );
};

const bootstrapPlugin = (
  engine, { help = 'help', helpDescription = 'Display this help menu', consoleCommands }
) => ({ dispatch }) => {
  bootstrap(engine)({
    dispatch,
    consoleCommands: {
      [help]: { description: helpDescription, command: logHelp },
      ...consoleCommands
    }
  });
};

const bootstrap = engine => ({ dispatch, consoleCommands }) => {
  if (consoleCommands) {
    dispatch(doBootstrap(consoleCommands));
    R.forEachObjIndexed(
      ({ command }, key) => {
        window[key] = () => command({ state: engine.getStore().getState(), dispatch });
      },
      consoleCommands
    );
  }
};
