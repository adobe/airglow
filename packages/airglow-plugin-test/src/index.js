/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { REDUCER } from 'airglow';
import { slice } from '@airglow/reducers';
import { last } from 'ramda';
import reducer from './reducer';

const STORE_KEY = '_test_dispatches';

export default function testPlugin() {
  return (engine) => {
    engine.plugin(REDUCER, slice(STORE_KEY).with(reducer));
  };
}

export const lastDispatch = state => last(state[STORE_KEY]);
export const dispatchList = state => state[STORE_KEY];
