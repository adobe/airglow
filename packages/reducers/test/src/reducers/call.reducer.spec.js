/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { assoc } from 'ramda';
import { call } from '../../../src/index';

let reducer;
let reducerB;
const args = [
  { pizza: 'sausage' },
  { type: 'changePizza' },
  { fullState: { food: { pizza: 'sausage' } } }
];

let goodVal;
let badVal;

describe('CallReducer', () => {
  beforeEach(() => {
    reducer = sinon.stub().callsFake(assoc('pizza', 'cheese'));
    reducerB = sinon.stub().callsFake(assoc('iceCream', 'rocky road'));

    goodVal = sinon.stub().returns(true);
    badVal = sinon.stub().returns(false);
  });
  it('should return the results of the provided function', function () {
    expect(call(reducer)(...args)).toMatchSnapshot();
  });
  it('should provide the reducer with the state and action', function () {
    call(reducer)(...args);
    expect(reducer.firstCall.args).toMatchSnapshot();
  });
  it('should support multiple reducers', function () {
    call(reducer, reducerB)(...args);
    expect(call(reducer, reducerB)(...args)).toMatchSnapshot();
  });
  it('should call if the for is a match', function () {
    const myCall = call(reducer).for('changePizza');
    myCall(...args);
    expect(reducer.callCount).toBe(1);
  });
  it('should not call if the for is not match', function () {
    const myCall = call(reducer).for('changePizza');
    myCall({ pizza: 'sausage' }, { type: 'addTopping' });
    expect(reducer.callCount).toBe(0);
  });
  it('should handle multiple fors', function () {
    const myCall = call(reducer).for('changePizza', 'addTopping');
    myCall({ pizza: 'sausage' }, { type: 'addTopping' });
    expect(reducer.callCount).toBe(1);
  });
  it('should fail if a validator fails', function () {
    const myCall = call(reducer).when(goodVal, badVal);
    myCall(...args);
    expect(reducer.callCount).toBe(0);
  });
  it('should succeed if all validators work', function () {
    const myCall = call(reducer).when(goodVal);
    myCall(...args);
    expect(reducer.callCount).toBe(1);
  });
  it('should pass the state and action to the validators', function () {
    const myCall = call(reducer).when(goodVal);
    myCall(...args);
    expect(goodVal.firstCall.args).toMatchSnapshot();
  });
  it('should provide the config for others to use', function () {
    const myCall = call(reducer).for('changePizza');
    expect(myCall.config).toMatchSnapshot();
  });
});
