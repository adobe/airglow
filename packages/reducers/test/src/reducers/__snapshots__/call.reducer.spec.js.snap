// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallReducer should pass the state and action to the validators 1`] = `
Array [
  Object {
    "pizza": "sausage",
  },
  Object {
    "type": "changePizza",
  },
]
`;

exports[`CallReducer should provide the config for others to use 1`] = `
Object {
  "fors": Array [
    "changePizza",
  ],
  "reducers": Array [
    [Function],
  ],
}
`;

exports[`CallReducer should provide the reducer with the state and action 1`] = `
Array [
  Object {
    "pizza": "sausage",
  },
  Object {
    "type": "changePizza",
  },
  Object {
    "fullState": Object {
      "food": Object {
        "pizza": "sausage",
      },
    },
    "reducers": Array [
      [Function],
    ],
  },
]
`;

exports[`CallReducer should return the results of the provided function 1`] = `
Object {
  "pizza": "cheese",
}
`;

exports[`CallReducer should support multiple reducers 1`] = `
Object {
  "iceCream": "rocky road",
  "pizza": "cheese",
}
`;
